File: internal/tiketkai/tiketkai.go
package tiketkai

import (
	"bytes"
	"context"
	"crypto/aes"
	"crypto/cipher"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"io"
	"log/slog"
	"net/http"
	"strings"
	"time"

	"tiket-kereta-notifier/internal/common"
)

const (
	APIUrl = "https://sc-microservice-tiketkai.bmsecure.id/train/search"
	Key    = "78455d8581f1fc41"
	IV     = "34f1cdf17d1aacb8"
)

// Provider implements common.Provider for TiketKai
type Provider struct {
	Logger        *slog.Logger
	Origin        string
	Destination   string
	Date          string // YYYY-MM-DD
	TrainName     string // Target train name filter
	CheckInterval time.Duration
}

// NewProvider creates a new TiketKai provider
func NewProvider(logger *slog.Logger, origin, dest, date, trainName string, interval time.Duration) *Provider {
	return &Provider{
		Logger:        logger,
		Origin:        origin,
		Destination:   dest,
		Date:          date,
		TrainName:     trainName,
		CheckInterval: interval,
	}
}

func (p *Provider) Name() string {
	return "TiketKai"
}

func (p *Provider) Search(ctx context.Context) ([]common.Train, error) {
	if p.Origin == "" || p.Destination == "" {
		return nil, fmt.Errorf("origin and destination required")
	}

	p.Logger.Info("Searching TiketKai", "origin", p.Origin, "dest", p.Destination, "date", p.Date)

	payload := map[string]interface{}{
		"app":         "TKAI",
		"via":         "mobile_web",
		"date":        p.Date,
		"destination": p.Destination,
		"origin":      p.Origin,
		"productCode": "WKAI",
		"deviceInfo": map[string]interface{}{
			"model":       "Windows NT 10.0",
			"versionCode": 10037,
			"versionName": "1.3.0",
		},
	}

	jsonBytes, err := json.Marshal(payload)
	if err != nil {
		return nil, err
	}

	encrypted, err := encryptAESBase64(string(jsonBytes), Key, IV)
	if err != nil {
		return nil, fmt.Errorf("encrypt failed: %w", err)
	}

	// Send encrypted data directly (not wrapped in JSON)
	req, err := http.NewRequestWithContext(ctx, "POST", APIUrl, bytes.NewBufferString(encrypted))
	if err != nil {
		return nil, err
	}
	req.Header.Add("accept", "application/json, text/plain, */*")
	req.Header.Add("accept-language", "en-US,en;q=0.9")
	req.Header.Add("content-type", "text/plain")
	req.Header.Add("origin", "https://m.tiketkai.com")
	req.Header.Add("priority", "u=1, i")
	req.Header.Add("referer", "https://m.tiketkai.com/")
	req.Header.Add("sec-ch-ua", "\"Not(A:Brand\";v=\"8\", \"Chromium\";v=\"144\", \"Microsoft Edge\";v=\"144\"")
	req.Header.Add("sec-ch-ua-mobile", "?0")
	req.Header.Add("sec-ch-ua-platform", "\"Windows\"")
	req.Header.Add("sec-fetch-dest", "empty")
	req.Header.Add("sec-fetch-mode", "cors")
	req.Header.Add("sec-fetch-site", "cross-site")
	req.Header.Add("user-agent", "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/144.0.0.0 Safari/537.36 Edg/144.0.0.0")

	client := &http.Client{Timeout: 45 * time.Second}
	resp, err := client.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("API error: %d", resp.StatusCode)
	}

	// Read full body for debugging
	bodyBytes, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("failed to read response: %w", err)
	}

	var result struct {
		RC   string          `json:"rc"`
		Data json.RawMessage `json:"data"` // Can be array or string
	}

	if err := json.Unmarshal(bodyBytes, &result); err != nil {
		p.Logger.Error("Failed to parse response", "body", string(bodyBytes))
		return nil, err
	}

	if result.RC != "00" {
		p.Logger.Error("TiketKai API error", "rc", result.RC, "body", string(bodyBytes))
		return nil, fmt.Errorf("API RC: %s", result.RC)
	}

	// Check if data is a string (error message) or array
	if len(result.Data) > 0 && result.Data[0] == '"' {
		// Data is a string, likely an error message
		var msg string
		json.Unmarshal(result.Data, &msg)
		return nil, fmt.Errorf("API message: %s", msg)
	}

	// Parse data as array
	var trainData []struct {
		TrainName     string `json:"trainName"`
		TrainNumber   string `json:"trainNumber"`
		DepartureTime string `json:"departureTime"`
		ArrivalTime   string `json:"arrivalTime"`
		Seats         []struct {
			Class        string      `json:"class"`
			Availability interface{} `json:"availability"`
			PriceAdult   interface{} `json:"priceAdult"`
		} `json:"seats"`
	}

	if err := json.Unmarshal(result.Data, &trainData); err != nil {
		return nil, fmt.Errorf("failed to parse train data: %w", err)
	}

	var trains []common.Train
	for _, tr := range trainData {
		seatsAvailable := "0"
		availStatus := "FULL"
		minPrice := ""

		for _, s := range tr.Seats {
			// Check availability
			isAvail := false
			seatCount := "0"

			switch v := s.Availability.(type) {
			case float64:
				if v > 0 {
					isAvail = true
					seatCount = fmt.Sprintf("%.0f", v)
				}
			case string:
				if v != "0" && v != "Habis" && v != "" {
					isAvail = true
					seatCount = v
				}
			}

			if isAvail {
				seatsAvailable = seatCount
				availStatus = "AVAILABLE"

				// Price
				price := "0"
				switch v := s.PriceAdult.(type) {
				case float64:
					price = fmt.Sprintf("%.0f", v)
				case string:
					price = v
				}
				minPrice = price // Use first available class price
				break            // Take the first available class
			}
		}

		if availStatus == "AVAILABLE" {
			trains = append(trains, common.Train{
				Name:          tr.TrainName,
				Class:         "ECO", // Simplified
				Price:         minPrice,
				DepartureTime: tr.DepartureTime,
				ArrivalTime:   tr.ArrivalTime,
				Availability:  availStatus,
				SeatsLeft:     seatsAvailable,
			})
		}
	}

	// Filter by TrainName if configured
	if p.TrainName != "" {
		var filtered []common.Train
		target := strings.ToLower(p.TrainName)
		for _, t := range trains {
			if strings.Contains(strings.ToLower(t.Name), target) {
				filtered = append(filtered, t)
			}
		}
		return filtered, nil
	}

	return trains, nil
}

// SearchAll returns all trains (no TrainName filter)
func (p *Provider) SearchAll(ctx context.Context) ([]common.Train, error) {
	// Temporarily clear TrainName to get all trains
	savedName := p.TrainName
	p.TrainName = ""
	defer func() { p.TrainName = savedName }()
	return p.Search(ctx)
}

func (p *Provider) StartScheduler(ctx context.Context, notifyFunc func(string)) {
	interval := p.CheckInterval
	if interval == 0 {
		interval = 1 * time.Minute
	}

	p.Logger.Info("Starting TiketKai Polling", "interval", interval, "target", p.TrainName)
	ticker := time.NewTicker(interval)
	defer ticker.Stop()

	for {
		select {
		case <-ctx.Done():
			return
		case <-ticker.C:
			trains, err := p.Search(ctx)
			if err != nil {
				p.Logger.Error("Poll failed", "error", err)
				continue
			}

			// Filter for trains with available seats only
			var availableTrains []common.Train
			for _, t := range trains {
				if t.SeatsLeft != "0" && t.SeatsLeft != "" {
					availableTrains = append(availableTrains, t)
				}
			}

			if len(availableTrains) > 0 {
				msg := fmt.Sprintf("🚂 Target Train Available! (%d found)\n", len(availableTrains))
				for _, t := range availableTrains {
					msg += fmt.Sprintf("- %s: %s seats (Rp%s)\n", t.Name, t.SeatsLeft, t.Price)
				}
				notifyFunc(msg)
			}
		}
	}
}

// Init handles legacy boilerplate compatibility (now empty)
func Init(ctx context.Context) error {
	return nil
}

// Helpers
func encryptAESBase64(plaintext string, key, iv string) (string, error) {
	block, err := aes.NewCipher([]byte(key))
	if err != nil {
		return "", err
	}

	// PKCS7 Padding
	bs := block.BlockSize()
	padding := bs - len(plaintext)%bs
	padtext := bytes.Repeat([]byte{byte(padding)}, padding)
	plaintextBytes := append([]byte(plaintext), padtext...)

	mode := cipher.NewCBCEncrypter(block, []byte(iv))
	ciphertext := make([]byte, len(plaintextBytes))
	mode.CryptBlocks(ciphertext, plaintextBytes)

	return base64.StdEncoding.EncodeToString(ciphertext), nil
}

File: notifier/providers/tiketkai.py
"""TiketKai.com provider — AES-CBC encrypted JSON API."""

from __future__ import annotations

import base64
import json
import logging

from cryptography.hazmat.primitives import padding
from cryptography.hazmat.primitives.ciphers import Cipher, algorithms, modes

from ..config import FlatTrainConfig
from ..models import Train
from .base import BaseProvider
from .http_util import make_client

API_URL = "https://sc-microservice-tiketkai.bmsecure.id/train/search"
AES_KEY = b"78455d8581f1fc41"
AES_IV = b"34f1cdf17d1aacb8"

HEADERS = {
    "accept": "application/json, text/plain, */*",
    "accept-language": "en-US,en;q=0.9",
    "content-type": "text/plain",
    "origin": "https://m.tiketkai.com",
    "priority": "u=1, i",
    "referer": "https://m.tiketkai.com/",
    "sec-ch-ua": '"Not:A-Brand";v="99", "Microsoft Edge";v="145", "Chromium";v="145"',
    "sec-ch-ua-mobile": "?0",
    "sec-ch-ua-platform": '"Windows"',
    "sec-fetch-dest": "empty",
    "sec-fetch-mode": "cors",
    "sec-fetch-site": "cross-site",
    "user-agent": (
        "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 "
        "(KHTML, like Gecko) Chrome/145.0.0.0 Safari/537.36 Edg/145.0.0.0"
    ),
}


def _encrypt_aes_base64(plaintext: str) -> str:
    padder = padding.PKCS7(algorithms.AES.block_size).padder()
    padded = padder.update(plaintext.encode("utf-8")) + padder.finalize()
    encryptor = Cipher(algorithms.AES(AES_KEY), modes.CBC(AES_IV)).encryptor()
    ciphertext = encryptor.update(padded) + encryptor.finalize()
    return base64.b64encode(ciphertext).decode("ascii")


class TiketKaiProvider(BaseProvider):
    provider_key = "tiketkai"
    include_class_in_notify = False
    default_interval = 60.0

    def __init__(self, logger: logging.Logger, flat: FlatTrainConfig, index: int) -> None:
        super().__init__(logger, flat, index)

    async def _fetch(self) -> tuple[list[Train], str]:
        if not self.origin or not self.destination:
            raise ValueError("origin and destination required")

        self.logger.info("Searching TiketKai origin=%s dest=%s date=%s", self.origin, self.destination, self.date)

        payload = {
            "app": "TKAI",
            "via": "mobile_web",
            "date": self.date,
            "destination": self.destination,
            "origin": self.origin,
            "productCode": "WKAI",
            "deviceInfo": {
                "model": "Windows NT 10.0",
                "versionCode": 10037,
                "versionName": "1.3.0",
            },
        }
        encrypted = _encrypt_aes_base64(json.dumps(payload))

        async with make_client(self.proxy_url, 120.0) as client:
            resp = await client.post(API_URL, content=encrypted, headers=HEADERS)

        if resp.status_code != 200:
            raise RuntimeError(f"TiketKai API HTTP {resp.status_code}: {resp.text[:200]}")

        body = resp.json()
        rc = body.get("rc")
        if rc != "00":
            raise RuntimeError(f"API RC: {rc}")

        data = body.get("data")
        if isinstance(data, str):
            raise RuntimeError(f"API message: {data}")

        trains: list[Train] = []
        for tr in data or []:
            seats_available = "0"
            avail_status = "FULL"
            min_price = 0
            for seat in tr.get("seats") or []:
                availability = seat.get("availability")
                is_avail = False
                seat_count = "0"
                if isinstance(availability, (int, float)):
                    if availability > 0:
                        is_avail = True
                        seat_count = str(int(availability))
                elif isinstance(availability, str) and availability not in ("0", "Habis", ""):
                    is_avail = True
                    seat_count = availability

                if is_avail:
                    seats_available = seat_count
                    avail_status = "AVAILABLE"
                    price = seat.get("priceAdult")
                    try:
                        min_price = int(float(price))
                    except (TypeError, ValueError):
                        min_price = 0
                    break

            trains.append(
                Train(
                    name=tr.get("trainName", ""),
                    train_class="ECO",
                    price=min_price,
                    departure_time=tr.get("departureTime", ""),
                    arrival_time=tr.get("arrivalTime", ""),
                    availability=avail_status,
                    seats_left=seats_available,
                )
            )

        return trains, "edge145"  # ponytail: keep in sync with HEADERS above
